### ISSUE TITLE
model performing well on the screenshot of the video , but not on the actual parent video

### ISSUE THREAD

**Issue by AbhinavJangra29** [ultralytics/ultralytics#8452]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

i trained a yolov8n on a gun dataset consisting of 70k+ images link:https://universe.roboflow.com/phillip-lavrador/70k-guns/dataset/5 . trained it for 40-50 epoch and got 96+ mAP score. now when i inference the model on videos it performs poorly and gives many false positives in the output video...but when i take a screenshot from that output video where the gun wasnt being detected and pass that image again to the model it detects the gun with high confidence. attached the media below

### Additional

![image](https://github.com/ultralytics/ultralytics/assets/107471490/c07662c4-dd56-45f2-b9ca-a5b01f7ec34f)

https://github.com/ultralytics/ultralytics/assets/107471490/8b399d4e-d754-469f-a61c-4bf62ef0d5d7

**Comment by glenn-jocher** [comment]
@AbhinavJangra29 hello! Thanks for reaching out with your question. It's great to hear about the high mAP score you achieved with your model on the gun dataset. The issue you're experiencing with video inference versus static images is intriguing. 

One possible explanation could be the difference in image quality or frame resolution between the static screenshots and the video frames. Videos, especially if they're compressed or of lower resolution, might not provide the same level of detail as a high-resolution screenshot, which could affect the model's detection performance.

Another factor could be motion blur in the video frames, which isn't present in static images. This can significantly impact the model's ability to detect objects accurately.

To mitigate these issues, you might consider:
- Ensuring the video is of high quality and resolution, similar to your training images.
- Applying image preprocessing techniques to enhance video frames before inference, such as sharpening filters to reduce motion blur.

Here's a quick example of how you could apply a sharpening filter using OpenCV in Python:

```python
import cv2
import numpy as np

# Load your video
cap = cv2.VideoCapture('path/to/your/video.mp4')

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    # Apply a sharpening kernel
    kernel = np.array([[0, -1, 0], [-1, 5,-1], [0, -1, 0]])
    sharpened_frame = cv2.filter2D(frame, -1, kernel)

    # Now you can pass `sharpened_frame` to your model for inference
    # model.predict(sharpened_frame)

cap.release()
```

This is just one approach, and the effectiveness can vary based on your specific video content. Experimenting with different preprocessing techniques or adjusting the model's confidence threshold for video inference might also yield better results. 

Hope this helps! Let us know if you have any more questions or if there's anything else we can assist you with. 😊

**Comment by AbhinavJangra29** [comment]
hey glen! thank you for giving your observations just to add some more things if i increase the epochs to 250-300 as well as the model size to s  and feed it with the same data but this time with added noise and annotations. and then downstream it to cctv footage gathered data do you think it is going to help? and thank you for the video sharpening tip will definitely add it! thanks!

**Comment by glenn-jocher** [comment]
@AbhinavJangra29 hey there! 😊 Increasing epochs to 250-300 and moving to a larger model size, like 's', can indeed help the model learn more complex patterns, especially when you introduce noise and augmentations to your training data. This approach often leads to better generalization on real-world data, such as CCTV footage. Just ensure your augmentations reflect the conditions seen in your target environment for the best results. The video sharpening tip should also complement these changes nicely. Happy experimenting, and feel free to reach out if you have more questions!

**Comment by AbhinavJangra29** [comment]
@glenn-jocher hey! just wanted you to know that with enough training and a larger architecture, its now able to detect object of interest at even somewhat a bad resolution like 360p and in confusing background . thank you ! love your work btw:)

**Comment by glenn-jocher** [comment]
Woohoo awesome!! Glad to hear everything is working better now 😃

**Comment by AbhinavJangra29** [comment]
@glenn-jocher hey! wanted to ask one more thing im working on a academic project for which i want to deploy this model and infer it on incoming video streams from cctvs. i read about the rtsp source that we can give  to the model directly and the model will infer on it, i tried it and the model starting giving its prediction in terminal. cant we see the rtsp stream with bounding boxes after inferred by the model?

**Comment by glenn-jocher** [comment]
@AbhinavJangra29 hey there! 😊 Absolutely, you can visualize the RTSP stream with bounding boxes overlaid on the detected objects. After receiving predictions from the model, you can use OpenCV to draw the bounding boxes and display the stream. Here's a quick example:

```python
import cv2
from ultralytics import YOLO

# Load your model
model = YOLO('yolov8n.pt')

# Connect to the RTSP stream
cap = cv2.VideoCapture('rtsp://your_rtsp_stream_link')

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    # Get predictions
    results = model(frame)

    # Draw bounding boxes and labels on the frame
    annotated_frame = results[0].plot()

    # Display the frame
    cv2.imshow('YOLOv8 RTSP Stream', annotated_frame)
    if cv2.waitKey(1) == ord('q'):  # Press 'q' to quit
        break

cap.release()
cv2.destroyAllWindows()
```

This code will open a window showing your RTSP stream with bounding boxes and labels. Just replace `'rtsp://your_rtsp_stream_link'` with your actual RTSP stream URL. Happy coding, and best of luck with your academic project! 🚀

**Comment by AbhinavJangra29** [comment]
hey glenn! while making an api for it i wanted to do one more task that is to save the frames with bounding boxes when it detected pistol class, but i am confused about the result attirbutes that it takes to plot the boxes. 
```
from fastapi import FastAPI, UploadFile, File, HTTPException
from pydantic import BaseModel
from ultralytics import YOLO
import os

my_app = FastAPI()
upload_dir = "uploads"
model = YOLO("300epochs.pt")
save_dir = "pistol_frames"  # New directory to save frames with detections

# Ensure that the upload directory and save directory exist
os.makedirs(upload_dir, exist_ok=True)
os.makedirs(save_dir, exist_ok=True)  # Create save directory


class RTSPRequest(BaseModel):
    url: str


@my_app.post("/detect-rtsp/")
async def detect_objects(rtsp_request: RTSPRequest):
    try:
        results = model(rtsp_request.url, stream=True)
        saved_frames = 0
        for result in results:
            if result.pandas().xyxy[0]['name'] == "pistol":  # Check for "pistol" class
                frame_path = os.path.join(save_dir, f"{result.frame}_pistol.jpg")
                result.save(frame_path)  # Save frame with pistol detection
                saved_frames += 1
        return {"detections": saved_frames, "message": f"{saved_frames} frames saved."}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@my_app.post("/detect-video/")
async def detect_objects(video_file: UploadFile = File(...)):
    try:
        # Save the uploaded video file
        file_path = os.path.join(upload_dir, video_file.filename)
        with open(file_path, "wb") as buffer:
            buffer.write(video_file.file.read())

        # Run inference on the video file
        results = model(file_path, stream=False)

        # Process the detection results
        saved_frames = 0
        for result in results.pandas().xyxy:  # Iterate through detections in pandas format
            if result['name'] == "pistol":
                frame_path = os.path.join(save_dir, f"{result['frame']}_pistol.jpg")  # Use frame number from results
                # Extract and save frame (code depends on YOLO library, potentially using result.vis.render())
                saved_frames += 1
        # You can optionally return the number of saved frames here

        return {"message": f"Object detection completed successfully! {saved_frames} frames saved."}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

```

when i call this api using postman i can successfully the see the model response on each frame but it is not able to save the output frames with bounding boxes around the object. please help!

**Comment by AbhinavJangra29** [comment]
what i want it to be like is - 
i give it the rtsp links
and it shows me a live output stream of the video with bounding boxes in it.

**Comment by glenn-jocher** [comment]
Hey there! 😊 It sounds like you're looking to serve real-time video streams with bounding boxes through FastAPI. For real-time streaming within a web framework like FastAPI, you might consider using WebSocket for continuous data flow. Unfortunately, FastAPI and similar frameworks don't directly support real-time video streaming with modifications like drawing bounding boxes on frames.

A workaround could involve sending the video stream to the client-side (browser) and using JavaScript to render the video while receiving bounding box coordinates via WebSocket. Then, you can draw the boxes on the video in the browser.

Here's a very simplified example of how you might set up the WebSocket endpoint in FastAPI:

```python
from fastapi import FastAPI, WebSocket

app = FastAPI()
 @AbhinavJangra29.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        # Process your stream here and send back the bounding box data
        await websocket.send_text(f"Message text was: {data}")
```

On the client side, you would connect to this WebSocket, send the video frames (or URLs, depending on your setup), and receive the coordinates for drawing the bounding boxes.

This approach requires handling video processing and bounding box rendering separately, which can be complex but allows for real-time streaming within the limitations of HTTP-based web frameworks.

For displaying the live output stream with bounding boxes, you might need a more complex setup involving client-side video rendering and server-side detection logic. I hope this points you in the right direction! 🚀

**Comment by AbhinavJangra29** [comment]
https://github.com/ultralytics/ultralytics/assets/107471490/b7c18c26-6567-45d2-a033-0386d31afa5b



hey @glenn-jocher ! didnt quiet get what is wrong with the script , i have attached the media.

also the code i am using is:
```
import cv2
import streamlit as st
from PIL import Image
from ultralytics import YOLO
import tempfile
import os
#import time

def main():
    st.title("Gun Detection")

    video_source_option = st.radio("Select Video Source:", ("Video File", "RTSP Stream", "Webcam"))

    if video_source_option == "Video File":
        video_file = st.file_uploader("Upload Video", type=["mp4", "avi"])
        if video_file is not None:
            # Create a temporary file to store the uploaded video
            with tempfile.NamedTemporaryFile(delete=False) as temp_file:
                temp_file.write(video_file.read())
                temp_file_path = temp_file.name
            detect_objects(temp_file_path)
            # Remove the temporary file after processing
            os.unlink(temp_file_path)

    elif video_source_option == "RTSP Stream":
        rtsp_link = st.text_input("Enter RTSP Link:")
        if st.button("Start Detection"):
            detect_objects(rtsp_link)

    elif video_source_option == "Webcam":
        detect_objects(0)

def detect_objects(video_source):
    yolo_model = YOLO('300epochs.pt')
    cap = cv2.VideoCapture(video_source)
    placeholder = st.empty()
    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        # Get predictions
        results = yolo_model(frame)

        # Draw bounding boxes and labels on the frame
        annotated_frame = results[0].plot()

        # Convert the annotated frame to RGB (Streamlit uses RGB)
        annotated_frame_rgb = cv2.cvtColor(annotated_frame, cv2.COLOR_BGR2RGB)

        # Convert the frame to Image
        img_pil = Image.fromarray(annotated_frame_rgb)

        # Display the frame
        placeholder.image(img_pil, use_column_width=True)
        #time.sleep(0.1)

if __name__ == '__main__':
    main()


```

**Comment by glenn-jocher** [comment]
Hey @AbhinavJangra29! 👋 It looks like your script is well-structured for integrating YOLOv8 with Streamlit for gun detection across different video sources. If you're encountering issues, it might be related to how the frames are being processed and displayed in real-time. 

One thing to check is the `results[0].plot()` method. Ensure it's correctly generating the annotated frames. If the issue persists, you might want to directly use OpenCV functions to draw the bounding boxes based on `results.pandas().xyxy[0]` data, which gives you more control over the visualization.

Here's a quick snippet on how you could draw bounding boxes manually:

```python
for i, det in results.pandas().xyxy[0].iterrows():  # Loop through detections
    cv2.rectangle(frame, 
                  (int(det['xmin']), int(det['ymin'])), 
                  (int(det['xmax']), int(det['ymax'])), 
                  (255, 0, 0), 2)  # Draw rectangle
    cv2.putText(frame, 
                f"{det['name']} {det['confidence']:.2f}", 
                (int(det['xmin']), int(det['ymin']-10)), 
                cv2.FONT_HERSHEY_SIMPLEX, 
                0.9, (255, 0, 0), 2)  # Put label
```

Replace the `results[0].plot()` line with the above code to manually draw bounding boxes and labels on `frame` before converting it to an RGB image and displaying it with Streamlit.

Hope this helps! Let me know if you have any more questions. 😊
